src/libertem/api.py

```diff
@@ -158,34 +183,94 @@ def make_with(cls, executor_spec: ExecutorSpecType, *args, **kwargs) -> 'Context
+        limited_execs = ('inline', 'synchronous', 'dask-integration', 'delayed')
+        cannot_cpus = executor_spec in limited_execs
+        if cpu_spec and cannot_cpus:
+            raise NotImplementedError(f'Executor type {executor_spec} does not support '
+                                      'specifying CPU workers at this time')
+        cannot_gpus = executor_spec in limited_execs + ('threads',)
+        if gpu_spec and cannot_gpus:
+            raise NotImplementedError(f'Executor type {executor_spec} does not support '
+                                      'specifying GPU workers at this time')
+        # Delay import here to avoid cupy import overhead
+        if has_spec and executor_spec in ('dask', 'dask-make-default', 'pipelined'):
+            from libertem.utils.devices import detect
+            spec_args = detect()
+            if cpu_spec:
+                spec_args['cpus'] = cpus
+            if gpu_spec:
+                if len(spec_args['cudas']) == 0 and gpus:
+                    raise ValueError('Cannot specify GPU workers as no GPUs detected')
+                spec_args['cudas'] = gpus
+        spec = None
+        if has_spec and executor_spec in ('dask', 'dask-make-default'):
+            spec = cluster_spec(**spec_args)
+
         executor: JobExecutor
         if executor_spec in ('synchronous', 'inline'):
             executor = InlineJobExecutor()
         elif executor_spec == 'threads':
-            executor = ConcurrentJobExecutor.make_local()
+            n_threads = cpus
+            try:
+                n_threads = len(n_threads)
+            except TypeError:
+                pass
+            executor = ConcurrentJobExecutor.make_local(n_threads=n_threads)
+        elif executor_spec == 'dask':
+            executor = DaskJobExecutor.make_local(spec=spec)
         elif executor_spec == 'dask-integration':
             executor = get_dask_integration_executor()
         elif executor_spec == 'dask-make-default':
-            executor = DaskJobExecutor.make_local(client_kwargs={"set_as_default": True})
+            executor = DaskJobExecutor.make_local(
+                spec=spec,
+                client_kwargs={"set_as_default": True}
+            )
         elif executor_spec == 'delayed':
             executor = DelayedJobExecutor()
         elif executor_spec == 'pipelined':
-            executor = PipelinedExecutor()
+            if has_spec:
+                spec = PipelinedExecutor.make_spec(**spec_args)
+                executor = PipelinedExecutor(spec=spec)
+            else:
+                executor = PipelinedExecutor.make_local()
         else:
             raise ValueError(
                 f'Argument `executor_spec` is {executor_spec}. Allowed are '
-                f'synchronous", "inline", "threads", "dask-integration", '
-                f'"dask-make-default" or "pipelined".'
+                f'synchronous", "inline", "threads", "dask", "dask-integration",'
```

Comment:
Missing a double-quote, btw:

```suggestion
                f'"synchronous", "inline", "threads", "dask", "dask-integration",'
```